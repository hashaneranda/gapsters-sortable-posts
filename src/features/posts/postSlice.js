import { createSlice } from "@reduxjs/toolkit";

/**
 * Swap the elements in an array at indexes x and y.
 *
 * @param (a) The array.
 * @param (x) The index of the first element to swap.
 * @param (y) The index of the second element to swap.
 * @return {Array} The input array with the elements swapped.
 */
const swapArrayElements = (a, x, y) => {
  console.log("initial array --", a, x);
  if (a.length === 1) return a;
  a.splice(y, 1, a.splice(x, 1, a[y])[0]);

  console.log("new array --", a);
  return a;
};

export const initialState = {
  postList: {
    loading: false,
    data: null,
    error: null,
  },
  sortArray: [],
};

const createdSlice = createSlice({
  name: "posts",
  initialState,
  reducers: {
    fetchPost(state, { payload }) {
      return {
        ...state,
        postList: { ...state.postList, loading: true },
      };
    },
    fetchPostSuccess(state, { response }) {
      return {
        ...state,
        postList: {
          ...state.postList,
          loading: false,
          data: response.data ? response.data.slice(0, 5) : response.data,
          error: null,
        },
      };
    },
    fetchPostError(state, { response }) {
      return {
        ...state,
        postList: {
          ...state.postList,
          loading: false,
          data: null,
          error: response
            ? response.data
              ? response.data.message
              : response.data
            : null,
        },
      };
    },
    moveUpPost(state, { payload }) {
      return {
        ...state,
        postList: {
          ...state.postList,
          loading: false,

          data: payload,
          error: null,
        },
      };
    },
    moveDownPost(state, { payload }) {
      return {
        ...state,
        postList: {
          ...state.postList,
          loading: false,
          data: payload,
          error: null,
        },
      };
    },
  },
});

const { actions, reducer } = createdSlice;

export const {
  fetchPost,
  fetchPostSuccess,
  fetchPostError,
  moveUpPost,
  moveDownPost,
} = actions;

export default reducer;
