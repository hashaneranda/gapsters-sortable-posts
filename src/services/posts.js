/**
 * Author : Hashan Eranda Jayalath
 * Copyrights: Hashan Eranda Jayalath
 * Version:
 * Description: posts service
 * Date: 02-10-2020
 */

import axios from "utils/API";

/**
 *Create File
 *
 * @param {*} request payload
 * @returns response
 */
export const fetchPosts = async () => {
  let apiRes = null;

  console.log("fetching");

  try {
    apiRes = await axios.get("/posts");
  } catch (err) {
    apiRes = err;
  } finally {
    console.log("posts", apiRes);
    return apiRes;
  }

  //   return [
  //     {
  //       userId: 1,
  //       id: 1,
  //       title:
  //         "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
  //       body:
  //         "quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto",
  //     },
  //     {
  //       userId: 1,
  //       id: 2,
  //       title: "qui est esse",
  //       body:
  //         "est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla",
  //     },
  //   ];
};
