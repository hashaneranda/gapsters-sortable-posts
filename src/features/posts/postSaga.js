import { put, call } from "redux-saga/effects";

//services
import * as service from "services/posts";

//actions
import * as actions from "./postSlice";

export function* fetchPostList({ payload }) {
  let response = null;

  try {
    response = yield call(service.fetchPosts, payload);

    if (response && response.status === 200) {
      yield put({
        type: actions.fetchPostSuccess.type,
        response,
      });
    } else {
      yield put({ type: actions.fetchPostError.type, response });
    }
  } catch (err) {
    yield put({ type: actions.fetchPostError.type, response });
  }
}
