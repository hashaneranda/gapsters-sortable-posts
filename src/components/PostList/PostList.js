import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useTransition, a } from "react-spring";

//compoenents
import Post from "components/Post/Post";

//redux actions
import { fetchPost, moveUpPost, moveDownPost } from "features/posts/postSlice";

//utils
import { swapArrayElements } from "./utils/ArrayHelpers";

const PostList = () => {
  const dispatch = useDispatch();
  const postList = useSelector((state) => state.posts.postList);

  const height = 50;
  //   const transitions = useTransition(
  //     rows.map((data, i) => ({ ...data, y: i * height })),
  //     d => d.name,
  //     {
  //       from: { position: "absolute", opacity: 0 },
  //       leave: { height: 0, opacity: 0 },
  //       enter: ({ y }) => ({ y, opacity: 1 }),
  //       update: ({ y }) => ({ y })
  //     }
  //   );

  useEffect(() => {
    dispatch(fetchPost());
  }, []);

  /**
   * Handle Post button click event
   *
   * @param {string} id
   * @param {string} action
   */
  const handleButtonClick = (id, action) => {
    console.log("action licked", action, " on id: ", id);

    if (action === "up") {
      const newList = swapArrayElements([...postList.data], id - 1, id);
      dispatch(moveUpPost(newList));
    } else if (action === "down") {
      const newList = swapArrayElements([...postList.data], id, id + 1);
      dispatch(moveDownPost(newList));
    }
  };

  return (
    <div data-testid="post-list" className="container mx-auto">
      <h1 className="text-2xl leading-loose">Sortable Post List</h1>

      {postList.data &&
        postList.data.map((data, index) => (
          <Post
            key={index}
            id={data.id}
            isStart={index === 0}
            isEnd={index === postList.data.length - 1}
            buttonUp={() => handleButtonClick(index, "up")}
            buttonDown={() => handleButtonClick(index, "down")}
          />
        ))}
    </div>
  );
};

export default PostList;
