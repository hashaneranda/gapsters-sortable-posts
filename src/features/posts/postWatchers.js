import { takeLatest, takeEvery } from "redux-saga/effects";

//actions
import * as postSlice from "features/posts/postSlice";

//saga
import * as postSaga from "features/posts/postSaga";

export default function* watchers() {
  //payment
  yield takeLatest(postSlice.fetchPost.type, postSaga.fetchPostList);
}
