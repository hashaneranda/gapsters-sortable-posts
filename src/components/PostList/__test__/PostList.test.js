import React from "react";

import { configureStore } from "@reduxjs/toolkit";

import { Provider } from "react-redux";
import { render, cleanup, fireEvent } from "@testing-library/react";

import { initialState } from "features/posts/postSlice";
import posts from "features/posts/postSlice";
import PostList from "../PostList";

const renderWithRedux = (
  component,
  {
    initialState,
    store = configureStore({
      reducer: {
        posts,
      },
      initialState,
    }),
  } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
};

afterEach(cleanup);

it("checks initial state is equal to 0", () => {
  const { getByTestId } = renderWithRedux(<PostList />);
  expect(getByTestId("post-list")).toBeInTheDocument();
});

it("should take a snapshot", () => {
  const { asFragment } = renderWithRedux(<PostList />);
  expect(asFragment(<PostList />)).toMatchSnapshot();
});
