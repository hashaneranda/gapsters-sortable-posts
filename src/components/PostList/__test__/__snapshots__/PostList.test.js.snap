// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto"
    data-testid="post-list"
  >
    <h1
      class="text-2xl leading-loose"
    >
      Sortable Post List
    </h1>
  </div>
</DocumentFragment>
`;
